numpy/core/src/multiarray/ctors.c

```diff
@@ -2069,6 +2069,50 @@ PyArray_FromStructInterface(PyObject *input)
     return NULL;
 }
 
+/*
+ * Checks if the object in descr is the default 'descr' member for the
+ * __array_interface__ dictionary with 'typestr' member typestr.
+ */
+NPY_NO_EXPORT int
+_is_default_descr(PyObject *descr, PyObject *typestr) {
+    PyObject *tuple, *name, *typestr2;
+#if defined(NPY_PY3K)
+    PyObject *tmp;
+#endif
+    int ret = 0;
+
+    if (!PyList_Check(descr) || PyList_GET_SIZE(descr) != 1) {
+        return 0;
+    }
+    tuple = PyList_GET_ITEM(descr, 0);
+    if (!(PyTuple_Check(tuple) && PyTuple_GET_SIZE(tuple) == 2)) {
+        return 0;
+    }
+    name = PyTuple_GET_ITEM(tuple, 0);
+    if (!(PyUString_Check(name) && PyUString_GET_SIZE(name) == 0)) {
+        return 0;
+    }
+    typestr2 = PyTuple_GET_ITEM(tuple, 1);
+#if defined(NPY_PY3K)
+    /* Allow unicode type strings */
+    if (PyUnicode_Check(typestr2)) {
+        tmp = PyUnicode_AsASCIIString(typestr2);
+        typestr2 = tmp;
+    }
+#endif
+    if (PyBytes_Check(typestr2) &&
+            PyObject_RichCompareBool(typestr, typestr2, Py_EQ)) {
+        ret = 1;
+    }
+#if defined(NPY_PY3K)
+    if (tmp == typestr2) {
+        Py_DECREF(tmp);
```

Comment:
Might want `Py_XDECREF` if tmp is NULL. Not sure how to handle the error case if one is possible.